### PR TITLE
set indexlabel invalid if create or rebuild failed

### PR DESCRIPTION
**Pull request by zhoney**
also check if indexlabel invalid when query by index

Change-Id: If890b3bd8c441b63642a2006479cf4ccca9ddf67

### CODE REVIEW

File: hugegraph-core/src/main/java/com/baidu/hugegraph/schema/builder/IndexLabelBuilder.java [hunk 1]
```diff
@@ -222,13 +222,19 @@ private boolean hasSameProperties(IndexLabel existedIndexLabel) {
...
+                // Async rebuild index
+                Id rebuildTask = this.rebuildIndex(indexLabel, removeTasks);
+                E.checkNotNull(rebuildTask, "rebuild-index task");
 
-            return new IndexLabel.CreatedIndexLabel(indexLabel, rebuildTask);
+                return new IndexLabel.CreatedIndexLabel(indexLabel,
+                                                        rebuildTask);
+            } catch (Throwable t) {
+                indexLabel.status(SchemaStatus.INVALID);
+                throw t;
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/schema/builder/IndexLabelBuilder.java:236]: rename t to e

**Review by Linary — DISMISSED** [review]

File: hugegraph-core/src/main/java/com/baidu/hugegraph/schema/builder/IndexLabelBuilder.java [hunk 2]
```diff
@@ -222,18 +222,18 @@ private boolean hasSameProperties(IndexLabel existedIndexLabel) {
...
                 Id rebuildTask = this.rebuildIndex(indexLabel, removeTasks);
                 E.checkNotNull(rebuildTask, "rebuild-index task");
 
                 return new IndexLabel.CreatedIndexLabel(indexLabel,
                                                         rebuildTask);
-            } catch (Throwable t) {
+            } catch (Throwable e) {
                 indexLabel.status(SchemaStatus.INVALID);
-                throw t;
+                this.graph().addIndexLabel(schemaLabel, indexLabel);
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/schema/builder/IndexLabelBuilder.java:235]: schemaTx.updateSchemaStatus(indexLabel, SchemaStatus.INVALID);

**Review by Linary — APPROVED** [review]

File: hugegraph-core/src/main/java/com/baidu/hugegraph/schema/builder/IndexLabelBuilder.java [hunk 3]
```diff
@@ -231,8 +231,7 @@ private boolean hasSameProperties(IndexLabel existedIndexLabel) {
                 return new IndexLabel.CreatedIndexLabel(indexLabel,
                                                         rebuildTask);
             } catch (Throwable e) {
-                indexLabel.status(SchemaStatus.INVALID);
-                this.graph().addIndexLabel(schemaLabel, indexLabel);
+                this.updateSchemaStatus(indexLabel, SchemaStatus.INVALID);
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/schema/builder/IndexLabelBuilder.java:234]: also consider removeSubIndex() at line 207

File: hugegraph-core/src/main/java/com/baidu/hugegraph/job/schema/IndexLabelRemoveCallable.java [hunk 4]
```diff
@@ -54,14 +54,20 @@ protected static void removeIndexLabel(HugeGraphParams graph, Id id) {
...
+                // Remove label from indexLabels of vertex or edge label
+                removeIndexLabelFromBaseLabel(schemaTx, indexLabel);
+                removeSchema(schemaTx, indexLabel);
+                // Should commit changes to backend store
+                // before release delete lock
+                graph.graph().tx().commit();
+            } catch (Throwable e) {
+                schemaTx.updateSchemaStatus(indexLabel, SchemaStatus.INVALID);
+                throw e;
+            }
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/job/schema/IndexLabelRemoveCallable.java:70]: should update status to INVALID  when failure to remove vertex/edge label?

File: hugegraph-core/src/main/java/com/baidu/hugegraph/job/schema/IndexLabelRemoveCallable.java [hunk 5]
```diff
@@ -54,14 +54,20 @@ protected static void removeIndexLabel(HugeGraphParams graph, Id id) {
...
-            graph.graph().tx().commit();
+            try {
+                // Remove index data
+                // TODO: use event to replace direct call
+                graphTx.removeIndex(indexLabel);
+                // Remove label from indexLabels of vertex or edge label
+                removeIndexLabelFromBaseLabel(schemaTx, indexLabel);
+                removeSchema(schemaTx, indexLabel);
+                // Should commit changes to backend store
+                // before release delete lock
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/job/schema/IndexLabelRemoveCallable.java:65]: use "/*"

File: hugegraph-core/src/main/java/com/baidu/hugegraph/job/schema/IndexLabelRemoveCallable.java (hunk 4 above)
Comment (zhoney) [reply hugegraph-core/src/main/java/com/baidu/hugegraph/job/schema/IndexLabelRemoveCallable.java:70]: no need. Fail to remove vertex/edge label will lead to partial delete of vertices/edges. vertex/edge label need to be still valid for remaining vertices/edges

**Review by houzhizhen — APPROVED** [review]

**Review by javeme — APPROVED** [review]

**Pull request merged**